The job **`x86_64-gnu-nopt`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/6428727006/job/17456432784) [(plain)](https://github.com/rust-lang-ci/rust/commit/d603d0feb223e7a4ad002d2fd02700ec3931e8f7/checks/17456432784/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [assembly] tests/assembly/libs/issue-115339-zip-arrays.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/libs/issue-115339-zip-arrays/issue-115339-zip-arrays.s" "/checkout/tests/assembly/libs/issue-115339-zip-arrays.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC" "--dump-input-context" "100"
--- stderr -------------------------------
/checkout/tests/assembly/libs/issue-115339-zip-arrays.rs:21:12: error: CHECK: expected string not found in input
/checkout/tests/assembly/libs/issue-115339-zip-arrays.rs:21:12: error: CHECK: expected string not found in input
 // CHECK: vorps
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/libs/issue-115339-zip-arrays/issue-115339-zip-arrays.s:1297:12: note: scanning from here
zip_arrays:
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/libs/issue-115339-zip-arrays/issue-115339-zip-arrays.s:1300:14: note: possible intended match here
 .cfi_def_cfa_offset 160

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/libs/issue-115339-zip-arrays/issue-115339-zip-arrays.s
Check file: /checkout/tests/assembly/libs/issue-115339-zip-arrays.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
            .
            .
            .
            .
         1197:  
         1198:  .section ".text._ZN94_$LT$core..slice..iter..IterMut$LT$T$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$24__iterator_get_unchecked17hfcdf675cc40b0b9fE","ax",@progbits 
         1199:  .globl _ZN94_$LT$core..slice..iter..IterMut$LT$T$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$24__iterator_get_unchecked17hfcdf675cc40b0b9fE 
         1200:  .p2align 4, 0x90 
         1201:  .type _ZN94_$LT$core..slice..iter..IterMut$LT$T$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$24__iterator_get_unchecked17hfcdf675cc40b0b9fE,@function 
         1202: _ZN94_$LT$core..slice..iter..IterMut$LT$T$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$24__iterator_get_unchecked17hfcdf675cc40b0b9fE: 
         1203:  .cfi_startproc 
         1204:  movq %rsi, %rax 
         1205:  addq (%rdi), %rax 
         1206:  retq 
         1207: .Lfunc_end25: 
         1208:  .size _ZN94_$LT$core..slice..iter..IterMut$LT$T$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$24__iterator_get_unchecked17hfcdf675cc40b0b9fE, .Lfunc_end25-_ZN94_$LT$core..slice..iter..IterMut$LT$T$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$24__iterator_get_unchecked17hfcdf675cc40b0b9fE 
         1209:  .cfi_endproc 
         1210:  
         1211:  .section ".text._ZN94_$LT$core..slice..iter..IterMut$LT$T$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$9size_hint17hb12900542b9b6a9eE","ax",@progbits 
         1212:  .globl _ZN94_$LT$core..slice..iter..IterMut$LT$T$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$9size_hint17hb12900542b9b6a9eE 
         1213:  .p2align 4, 0x90 
         1214:  .type _ZN94_$LT$core..slice..iter..IterMut$LT$T$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$9size_hint17hb12900542b9b6a9eE,@function 
         1215: _ZN94_$LT$core..slice..iter..IterMut$LT$T$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$9size_hint17hb12900542b9b6a9eE: 
         1216:  .cfi_startproc 
         1217:  subq $56, %rsp 
         1218:  .cfi_def_cfa_offset 64 
         1219:  movq %rsi, 8(%rsp) 
         1220:  movq %rdi, 16(%rsp) 
         1221:  movq %rdi, 24(%rsp) 
         1222:  xorl %eax, %eax 
         1223:  testb $1, %al 
         1224:  jne .LBB26_2 
         1225:  movq 8(%rsp), %rax 
         1226:  movq 8(%rax), %rdi 
         1227:  movq (%rax), %rsi 
         1228:  callq *_ZN4core3ptr8non_null16NonNull$LT$T$GT$7sub_ptr17hdd76a34969f26853E@GOTPCREL(%rip) 
         1229:  movq %rax, 32(%rsp) 
         1230:  jmp .LBB26_3 
         1231: .LBB26_2: 
         1232:  movq 8(%rsp), %rax 
         1233:  movq 8(%rax), %rax 
         1234:  movq %rax, 32(%rsp) 
         1235: .LBB26_3: 
         1236:  movq 24(%rsp), %rax 
         1237:  movq 16(%rsp), %rcx 
         1238:  movq 32(%rsp), %rdx 
         1239:  movq 32(%rsp), %rsi 
         1240:  movq %rsi, 48(%rsp) 
         1241:  movq $1, 40(%rsp) 
         1242:  movq %rdx, (%rcx) 
         1243:  movq 40(%rsp), %rsi 
         1244:  movq 48(%rsp), %rdx 
         1245:  movq %rsi, 8(%rcx) 
         1246:  movq %rdx, 16(%rcx) 
         1247:  addq $56, %rsp 
         1248:  .cfi_def_cfa_offset 8 
         1249:  retq 
         1250: .Lfunc_end26: 
         1251:  .size _ZN94_$LT$core..slice..iter..IterMut$LT$T$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$9size_hint17hb12900542b9b6a9eE, .Lfunc_end26-_ZN94_$LT$core..slice..iter..IterMut$LT$T$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$9size_hint17hb12900542b9b6a9eE 
         1252:  .cfi_endproc 
         1253:  
         1254:  .section ".text._ZN99_$LT$core..array..iter..IntoIter$LT$T$C$_$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$24__iterator_get_unchecked17h9e9435594ae797b0E","ax",@progbits 
         1255:  .globl _ZN99_$LT$core..array..iter..IntoIter$LT$T$C$_$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$24__iterator_get_unchecked17h9e9435594ae797b0E 
         1256:  .p2align 4, 0x90 
         1257:  .type _ZN99_$LT$core..array..iter..IntoIter$LT$T$C$_$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$24__iterator_get_unchecked17h9e9435594ae797b0E,@function 
         1258: _ZN99_$LT$core..array..iter..IntoIter$LT$T$C$_$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$24__iterator_get_unchecked17h9e9435594ae797b0E: 
         1259:  .cfi_startproc 
         1260:  pushq %rax 
         1261:  .cfi_def_cfa_offset 16 
         1262:  addq 32(%rdi), %rdi 
         1263:  addq %rsi, %rdi 
         1264:  callq *_ZN4core3ptr4read17h582f3a2f33710dc3E@GOTPCREL(%rip) 
         1265:  popq %rcx 
         1266:  .cfi_def_cfa_offset 8 
         1267:  retq 
         1268: .Lfunc_end27: 
         1269:  .size _ZN99_$LT$core..array..iter..IntoIter$LT$T$C$_$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$24__iterator_get_unchecked17h9e9435594ae797b0E, .Lfunc_end27-_ZN99_$LT$core..array..iter..IntoIter$LT$T$C$_$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$24__iterator_get_unchecked17h9e9435594ae797b0E 
         1270:  .cfi_endproc 
         1271:  
         1272:  .section ".text._ZN99_$LT$core..array..iter..IntoIter$LT$T$C$_$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$9size_hint17h57ed49009b8c8ad5E","ax",@progbits 
         1273:  .globl _ZN99_$LT$core..array..iter..IntoIter$LT$T$C$_$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$9size_hint17h57ed49009b8c8ad5E 
         1274:  .p2align 4, 0x90 
         1275:  .type _ZN99_$LT$core..array..iter..IntoIter$LT$T$C$_$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$9size_hint17h57ed49009b8c8ad5E,@function 
         1276: _ZN99_$LT$core..array..iter..IntoIter$LT$T$C$_$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$9size_hint17h57ed49009b8c8ad5E: 
         1277:  .cfi_startproc 
         1278:  movq %rdi, %rax 
         1279:  movq 40(%rsi), %rcx 
         1280:  subq 32(%rsi), %rcx 
         1281:  movq %rcx, -8(%rsp) 
         1282:  movq $1, -16(%rsp) 
         1283:  movq %rcx, (%rdi) 
         1284:  movq -16(%rsp), %rdx 
         1285:  movq -8(%rsp), %rcx 
         1286:  movq %rdx, 8(%rdi) 
         1287:  movq %rcx, 16(%rdi) 
         1288:  retq 
         1289: .Lfunc_end28: 
         1290:  .size _ZN99_$LT$core..array..iter..IntoIter$LT$T$C$_$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$9size_hint17h57ed49009b8c8ad5E, .Lfunc_end28-_ZN99_$LT$core..array..iter..IntoIter$LT$T$C$_$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$9size_hint17h57ed49009b8c8ad5E 
         1291:  .cfi_endproc 
         1292:  
         1293:  .section .text.zip_arrays,"ax",@progbits 
         1294:  .globl zip_arrays 
         1295:  .p2align 4, 0x90 
         1296:  .type zip_arrays,@function 
         1297: zip_arrays: 
check:21'0                X error: no match found
         1298:  .cfi_startproc 
check:21'0     ~~~~~~~~~~~~~~~~
         1299:  subq $152, %rsp 
check:21'0     ~~~~~~~~~~~~~~~~~
         1300:  .cfi_def_cfa_offset 160 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
check:21'1                  ?            possible intended match
         1301:  movq %rsi, 8(%rsp) 
check:21'0     ~~~~~~~~~~~~~~~~~~~~
         1302:  movq %rdi, 16(%rsp) 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~
         1303:  movq %rdi, 24(%rsp) 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~
         1304:  vmovdqu (%rdx), %ymm0 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~
         1305:  vmovdqu %ymm0, 120(%rsp) 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
         1306:  leaq 32(%rsp), %rdi 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~
         1307:  leaq 120(%rsp), %rdx 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~
         1308:  vzeroupper 
check:21'0     ~~~~~~~~~~~~
         1309:  callq *_ZN4core4iter8adapters3zip3zip17hd0801a31ed1f985eE@GOTPCREL(%rip) 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1310:  leaq 32(%rsp), %rdi 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~
         1311:  callq *_ZN4core4iter6traits8iterator8Iterator8for_each17h23a3583b210da5f6E@GOTPCREL(%rip) 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1312:  movq 8(%rsp), %rsi 
check:21'0     ~~~~~~~~~~~~~~~~~~~~
         1313:  movq 16(%rsp), %rdi 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~
         1314:  movq 24(%rsp), %rax 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~
         1315:  vmovdqu (%rsi), %ymm0 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~
         1316:  vmovdqu %ymm0, (%rdi) 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~
         1317:  addq $152, %rsp 
check:21'0     ~~~~~~~~~~~~~~~~~
         1318:  .cfi_def_cfa_offset 8 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~
         1319:  vzeroupper 
check:21'0     ~~~~~~~~~~~~
         1320:  retq 
check:21'0     ~~~~~~
         1321: .Lfunc_end29: 
check:21'0     ~~~~~~~~~~~~~~
         1322:  .size zip_arrays, .Lfunc_end29-zip_arrays 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1323:  .cfi_endproc 
check:21'0     ~~~~~~~~~~~~~~
         1324:  
check:21'0     ~
         1325:  .section ".text._ZN23issue_115339_zip_arrays10zip_arrays28_$u7b$$u7b$closure$u7d$$u7d$17h3c2099504b822f7cE","ax",@progbits 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1326:  .p2align 4, 0x90 
check:21'0     ~~~~~~~~~~~~~~~~~~
         1327:  .type _ZN23issue_115339_zip_arrays10zip_arrays28_$u7b$$u7b$closure$u7d$$u7d$17h3c2099504b822f7cE,@function 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1328: _ZN23issue_115339_zip_arrays10zip_arrays28_$u7b$$u7b$closure$u7d$$u7d$17h3c2099504b822f7cE: 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1329:  .cfi_startproc 
check:21'0     ~~~~~~~~~~~~~~~~
         1330:  movb %dl, %al 
check:21'0     ~~~~~~~~~~~~~~~
         1331:  orb (%rsi), %al 
check:21'0     ~~~~~~~~~~~~~~~~~
         1332:  movb %al, (%rsi) 
check:21'0     ~~~~~~~~~~~~~~~~~~
         1333:  retq 
check:21'0     ~~~~~~
         1334: .Lfunc_end30: 
check:21'0     ~~~~~~~~~~~~~~
         1335:  .size _ZN23issue_115339_zip_arrays10zip_arrays28_$u7b$$u7b$closure$u7d$$u7d$17h3c2099504b822f7cE, .Lfunc_end30-_ZN23issue_115339_zip_arrays10zip_arrays28_$u7b$$u7b$closure$u7d$$u7d$17h3c2099504b822f7cE 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1336:  .cfi_endproc 
check:21'0     ~~~~~~~~~~~~~~
         1337:  
check:21'0     ~
         1338:  .type .L__unnamed_14,@object 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1339:  .section .rodata..L__unnamed_14,"a",@progbits 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1340: .L__unnamed_14: 
check:21'0     ~~~~~~~~~~~~~~~~
         1341:  .ascii "/rustc/FAKE_PREFIX/library/core/src/iter/adapters/zip.rs" 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1342:  .size .L__unnamed_14, 56 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
         1343:  
check:21'0     ~
         1344:  .type .L__unnamed_1,@object 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1345:  .section .data.rel.ro..L__unnamed_1,"aw",@progbits 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1346:  .p2align 3, 0x0 
check:21'0     ~~~~~~~~~~~~~~~~~
         1347: .L__unnamed_1: 
check:21'0     ~~~~~~~~~~~~~~~
         1348:  .quad .L__unnamed_14 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~
         1349:  .asciz "8\000\000\000\000\000\000\000\f\001\000\000\023\000\000" 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1350:  .size .L__unnamed_1, 24 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
         1351:  
check:21'0     ~
         1352:  .type str.0,@object 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~
         1353:  .section .rodata.str.0,"a",@progbits 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1354:  .p2align 4, 0x0 
check:21'0     ~~~~~~~~~~~~~~~~~
         1355: str.0: 
check:21'0     ~~~~~~~
         1356:  .ascii "attempt to add with overflow" 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1357:  .size str.0, 28 
check:21'0     ~~~~~~~~~~~~~~~~~
         1358:  
check:21'0     ~
         1359:  .type .L__unnamed_2,@object 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1360:  .section .rodata..L__unnamed_2,"a",@progbits 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1361:  .p2align 3, 0x0 
check:21'0     ~~~~~~~~~~~~~~~~~
         1362: .L__unnamed_2: 
check:21'0     ~~~~~~~~~~~~~~~
         1363:  .size .L__unnamed_2, 0 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~
         1364:  
check:21'0     ~
         1365:  .type .L__unnamed_15,@object 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1366:  .section .rodata..L__unnamed_15,"a",@progbits 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1367: .L__unnamed_15: 
check:21'0     ~~~~~~~~~~~~~~~~
         1368:  .ascii "invalid args" 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~
         1369:  .size .L__unnamed_15, 12 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
         1370:  
check:21'0     ~
         1371:  .type .L__unnamed_3,@object 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1372:  .section .data.rel.ro..L__unnamed_3,"aw",@progbits 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1373:  .p2align 3, 0x0 
check:21'0     ~~~~~~~~~~~~~~~~~
         1374: .L__unnamed_3: 
check:21'0     ~~~~~~~~~~~~~~~
         1375:  .quad .L__unnamed_15 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~
         1376:  .asciz "\f\000\000\000\000\000\000" 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1377:  .size .L__unnamed_3, 16 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
         1378:  
check:21'0     ~
         1379:  .type .L__unnamed_16,@object 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1380:  .section .rodata..L__unnamed_16,"a",@progbits 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1381: .L__unnamed_16: 
check:21'0     ~~~~~~~~~~~~~~~~
         1382:  .ascii "/rustc/FAKE_PREFIX/library/core/src/fmt/mod.rs" 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1383:  .size .L__unnamed_16, 46 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
         1384:  
check:21'0     ~
         1385:  .type .L__unnamed_4,@object 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1386:  .section .data.rel.ro..L__unnamed_4,"aw",@progbits 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1387:  .p2align 3, 0x0 
check:21'0     ~~~~~~~~~~~~~~~~~
         1388: .L__unnamed_4: 
check:21'0     ~~~~~~~~~~~~~~~
         1389:  .quad .L__unnamed_16 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~
         1390:  .asciz ".\000\000\000\000\000\000\000?\001\000\000\r\000\000" 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1391:  .size .L__unnamed_4, 24 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
         1392:  
check:21'0     ~
         1393:  .type .L__unnamed_5,@object 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1394:  .section .rodata..L__unnamed_5,"a",@progbits 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1395: .L__unnamed_5: 
check:21'0     ~~~~~~~~~~~~~~~
         1396:  .ascii "unsafe precondition(s) violated: ptr::read requires that the pointer argument is aligned and non-null" 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1397:  .size .L__unnamed_5, 101 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
         1398:  
check:21'0     ~
         1399:  .type .L__unnamed_6,@object 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         1400:  .section .rodata..L__unnamed_6,"a",@progbits 
check:21'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            .
            .
>>>>>>
------------------------------------------
```

</details>

--- a/library/core/src/array/iter.rs
+++ b/library/core/src/array/iter.rs
@@ -4,7 +4,7 @@ use crate::num::NonZeroUsize;
 use crate::{
     fmt,
     intrinsics::transmute_unchecked,
-    iter::{self, ExactSizeIterator, FusedIterator, TrustedLen},
+    iter::{self, ExactSizeIterator, FusedIterator, TrustedLen, TrustedRandomAccessNoCoerce},
     mem::MaybeUninit,
     ops::{IndexRange, Range},
     ptr,
@@ -293,6 +293,12 @@ impl<T, const N: usize> Iterator for IntoIter<T, N> {
 
         NonZeroUsize::new(remaining).map_or(Ok(()), Err)
     }
+
+    #[inline]
+    unsafe fn __iterator_get_unchecked(&mut self, idx: usize) -> Self::Item {
+        // SAFETY: The caller must provide an idx that is in bound of the remainder.
+        unsafe { self.data.as_ptr().add(self.alive.start()).add(idx).cast::<T>().read() }
+    }
 }
 
 #[stable(feature = "array_value_iter_impls", since = "1.40.0")]
@@ -373,6 +379,25 @@ impl<T, const N: usize> FusedIterator for IntoIter<T, N> {}
 // always decremented by 1 in those methods, but only if `Some(_)` is returned.
 #[stable(feature = "array_value_iter_impls", since = "1.40.0")]
 unsafe impl<T, const N: usize> TrustedLen for IntoIter<T, N> {}
+
+#[doc(hidden)]
+#[unstable(issue = "none", feature = "std_internals")]
+#[rustc_unsafe_specialization_marker]
+pub trait NonDrop {}
+
+// T: Copy as approximation for !Drop since get_unchecked does not advance self.alive
+// and thus we can't implement drop-handling
+#[unstable(issue = "none", feature = "std_internals")]
+impl<T: Copy> NonDrop for T {}
+
+#[doc(hidden)]
+#[unstable(issue = "none", feature = "std_internals")]
+unsafe impl<T, const N: usize> TrustedRandomAccessNoCoerce for IntoIter<T, N>
+where
+    T: NonDrop,
+{
+    const MAY_HAVE_SIDE_EFFECT: bool = false;
+}
 
 #[stable(feature = "array_value_iter_impls", since = "1.40.0")]
 impl<T: Clone, const N: usize> Clone for IntoIter<T, N> {

--- a/library/core/src/iter/adapters/zip.rs
+++ b/library/core/src/iter/adapters/zip.rs
@@ -95,6 +95,14 @@ where
     }
 
     #[inline]
+    fn fold<Acc, F>(self, init: Acc, f: F) -> Acc
+    where
+        F: FnMut(Acc, Self::Item) -> Acc,
+    {
+        ZipImpl::fold(self, init, f)
+    }
+
+    #[inline]
     unsafe fn __iterator_get_unchecked(&mut self, idx: usize) -> Self::Item
     where
         Self: TrustedRandomAccessNoCoerce,
@@ -129,6 +137,9 @@ trait ZipImpl<A, B> {
     where
         A: DoubleEndedIterator + ExactSizeIterator,
         B: DoubleEndedIterator + ExactSizeIterator;
+    fn fold<Acc, F>(self, init: Acc, f: F) -> Acc
+    where
+        F: FnMut(Acc, Self::Item) -> Acc;
     // This has the same safety requirements as `Iterator::__iterator_get_unchecked`
     unsafe fn get_unchecked(&mut self, idx: usize) -> <Self as Iterator>::Item
     where
@@ -228,6 +239,14 @@ where
     {
         unreachable!("Always specialized");
     }
+
+    #[inline]
+    default fn fold<Acc, F>(self, init: Acc, f: F) -> Acc
+    where
+        F: FnMut(Acc, Self::Item) -> Acc,
+    {
+        SpecFold::spec_fold(self, init, f)
+    }
 }
 
 #[doc(hidden)]
@@ -250,6 +269,24 @@ where
         // SAFETY: the caller must uphold the contract for
         // `Iterator::__iterator_get_unchecked`.
         unsafe { (self.a.__iterator_get_unchecked(idx), self.b.__iterator_get_unchecked(idx)) }
+    }
+
+    #[inline]
+    fn fold<Acc, F>(mut self, init: Acc, mut f: F) -> Acc
+    where
+        F: FnMut(Acc, Self::Item) -> Acc,
+    {
+        let mut accum = init;
+        let len = ZipImpl::size_hint(&self).0;
+        for i in 0..len {
+            // SAFETY: since Self: TrustedRandomAccessNoCoerce we can trust the size-hint to
+            // calculate the length and then use that to do unchecked iteration.
+            // fold consumes the iterator so we don't need to fixup any state.
+            unsafe {
+                accum = f(accum, self.get_unchecked(i));
+            }
+        }
+        accum
     }
 }
 
@@ -588,5 +625,58 @@ unsafe impl<I: Iterator + TrustedRandomAccessNoCoerce> SpecTrustedRandomAccess f
         // SAFETY: the caller must uphold the contract for
         // `Iterator::__iterator_get_unchecked`.
         unsafe { self.__iterator_get_unchecked(index) }
+    }
+}
+
+trait SpecFold: Iterator {
+    fn spec_fold<B, F>(self, init: B, f: F) -> B
+    where
+        Self: Sized,
+        F: FnMut(B, Self::Item) -> B;
+}
+
+impl<A: Iterator, B: Iterator> SpecFold for Zip<A, B> {
+    // Adapted from default impl from the Iterator trait
+    #[inline]
+    default fn spec_fold<Acc, F>(mut self, init: Acc, mut f: F) -> Acc
+    where
+        F: FnMut(Acc, Self::Item) -> Acc,
+    {
+        let mut accum = init;
+        while let Some(x) = ZipImpl::next(&mut self) {
+            accum = f(accum, x);
+        }
+        accum
+    }
+}
+
+impl<A: TrustedLen, B: TrustedLen> SpecFold for Zip<A, B> {
+    #[inline]
+    fn spec_fold<Acc, F>(mut self, init: Acc, mut f: F) -> Acc
+    where
+        F: FnMut(Acc, Self::Item) -> Acc,
+    {
+        let mut accum = init;
+        loop {
+            let (upper, more) = if let Some(upper) = ZipImpl::size_hint(&self).1 {
+                (upper, false)
+            } else {
+                // Per TrustedLen contract a None upper bound means more than usize::MAX items
+                (usize::MAX, true)
+            };
+
+            for _ in 0..upper {
+                let pair =
+                    // SAFETY: TrustedLen guarantees that at least `upper` many items are available
+                    // therefore we know they can't be None
+                    unsafe { (self.a.next().unwrap_unchecked(), self.b.next().unwrap_unchecked()) };
+                accum = f(accum, pair);
+            }
+
+            if !more {
+                break;
+            }
+        }
+        accum
     }
 }

--- a/library/core/tests/iter/adapters/zip.rs
+++ b/library/core/tests/iter/adapters/zip.rs
@@ -184,7 +184,11 @@ fn test_zip_nested_sideffectful() {
         let it = xs.iter_mut().map(|x| *x = 1).enumerate().zip(&ys);
         it.count();
     }
-    assert_eq!(&xs, &[1, 1, 1, 1, 1, 0]);
+    let length_aware =  &xs == &[1, 1, 1, 1, 0, 0];
+    let probe_first = &xs == &[1, 1, 1, 1, 1, 0];
+
+    // either implementation is valid according to zip documentation
+    assert!(length_aware || probe_first);
 }
 
 #[test]

--- a/tests/assembly/libs/issue-115339-zip-arrays.rs
+++ b/tests/assembly/libs/issue-115339-zip-arrays.rs
@@ -1,0 +1,25 @@
+// assembly-output: emit-asm
+// # zen3 previously exhibited odd vectorization
+// compile-flags: --crate-type=lib -Ctarget-cpu=znver3 -O
+// only-x86_64
+// ignore-sgx
+
+use std::iter;
+
+// previously this produced a long chain of
+// 56:  vpextrb $6, %xmm0, %ecx
+// 57:  orb %cl, 22(%rsi)
+// 58:  vpextrb $7, %xmm0, %ecx
+// 59:  orb %cl, 23(%rsi)
+// [...]
+
+// CHECK-LABEL: zip_arrays:
+#[no_mangle]
+pub fn zip_arrays(mut a: [u8; 32], b: [u8; 32]) -> [u8; 32] {
+    // CHECK-NOT: vpextrb
+    // CHECK-NOT: orb %cl
+    // CHECK: vorps
+    iter::zip(&mut a, b).for_each(|(a, b)| *a |= b);
+    // CHECK: retq
+    a
+}

